scripts/k8s-diff/README.md

```diff
@@ -0,0 +1,274 @@
+2. Validation
+    - During this phase, debug information collected during preprocessing is used to validate that all rules defined in the config file actually impact the final output.
+
+```mermaid
+flowchart LR
+    Input --> Patching
+    Patching --> Validation
+    Validation --> Output
+```
+
+### Usage
+
+```
+Usage of yaml-patch:
+  -input-dir value
+    	Input directory, can be specified multiple times - must have the same number of elements as output-dir
+  -output-dir value
+    	Output directory, can be specified multiple times - must have the same number of elements as input-dir
+  -output-template string
+    	Template used to generate output file names.
+  -rules value
+    	Rule file to load, can be specified multiple times
+```
+
+A typical invocation might look like this:
+
+```
+yaml-patch -input-dir input-dir -output-dir output-dir \
+     -rules renames.yml \
+     -rules ignored_fields.yml 
+```
+
+### Rule File Format
+
+Rule files can be specified multiple times via the `-rules` flag. Rules across all files are collected and run in the following order
+
+1. Ignore rules from all files, in the order specified
+3. Patch rules from all files, in the order specified
+
+Below is an example rule file that ignores all objects with `kind: Secret` and ignores annotations across all objects.
+
+```
+---
+ignore_rules:
+- name: "Ignore all secrets"
+  match:
+  - op: test
+    path: /kind
+    value: Secret
+
+patch_rules:
+- remove_field: /metadata/annotations
+```
+
+All rules have a `name` and `match` field. See below for detailed notes about available fields on each rule.
+
+
+#### Ignore Rules
+
+Ignore rules are used to completely remove any object from the output.
+
+Ignore rules have the following structure:
+
+```
+- name: string
+  match: []JsonPatchOperation
+```
+
+- `name` is meant for documentation only. It is used in the program output to communicate issues to the user.
+- `match` is a [JsonPatch](#note-about-jsonpatch) value. If the patch can be successfully applied to a given object, then that object is said to "match" the ignore rule. Any objects matching the ignore rule are filtered from the output entirely. An empty `match` section will match every object.
+
+#### Patch Rules
+
+Patch rules are used to modify objects before the final diff is computed. 
+This is most commonly used to rename objects or alter their fields such that the rendered diff only contains relevant information.
+
+Patch rules have the following structure:
+
+```
+- name: string
+  match: []JsonPatchOperation (optional)
+  steps: []JsonPatchOperation
+  remove_field: jsonpointer (optional), (e.g. /metadata/labels/name)
+  rename_object: (optional)
+    from: string, name to replace in /metadata/name
+    to: string, target name
+  rename_field: (optional)
+    from: jsonpointer (optional), (e.g. /metadata/labels/name)
+    to: jsonpointer (optional), (e.g. /metadata/labels/name)
+  <jsonpointer>: []any (optional)
```

Comment:
having matchers in two places is slightly confusing, wdyt about something like moving the jsonpointer matchers in the `match` block?

```yaml
- name: string
  match: 
    jsonpointer: any (optional)
    jsonpointer: any (optional)
    ...
  # the rest is same as you have it
  #steps: []JsonPatchOperation
  #remove_field: jsonpointer (optional), (e.g. /metadata/labels/name)
  #rename_object: (optional)
  #  from: string, name to replace in /metadata/name
  #  to: string, target name
  #rename_field: (optional)
  #  from: jsonpointer (optional), (e.g. /metadata/labels/name)
  #  to: jsonpointer (optional), (e.g. /metadata/labels/name)
```

and optionally making the key value of the jsonpointer a single value, not an array, because I didn't see it being used as an array

I think this will simplify reading the matchers. It will improve the visual distinction between `match` items and the `steps` items, which now look identical and their parent blocks (match and steps) have the same number of letters

Reply:
I love this idea! Since `match` is an array, it might need to look just a little different:

```
- name: string
  match:
  - jsonpointer: any
  - op: test
    ...
```

This is slightly longer, but I think it's way clearer what's going on, and length of rule file is definitely not a goal to optimize for over understanding.

Reply:
Also, the multiple-value form is used a bunch in the latest version of the rules. It helps to DRY the config quite a bit in certain sections. I'll sync that to this branch in a bit.